CLogger/include/CLogger/CLogger.h

```diff
@@ -0,0 +1,314 @@
+			std::ostream& stream) const;
+		template<typename TupleType, size_t... Size>
+		std::ostream& PrintParams(const TupleType& tuple,
+			std::index_sequence<Size...>, std::ostream& stream) const;
+
+		void PrintToAllStreams(const std::string& info) const;
+		std::ostream& PrintLogInfo(const std::string& info,
+			std::ostream& stream) const;
+	};
+
+	/// <summary>
+	///		Clears all <c>ELogConfig</c> that was stored at <c>CLogger</c>
+	///		And than adds endless amount of <c>ELogConfig</c>. Order is important.
+	///		Duplicates will be ignored.
+	/// </summary>
+	/// <typeparam name="...Args">
+	///		Variadic template types. Waits just for <c>ELogConfig</c>.
+	///	</typeparam>
+	/// <param name="args">
+	///		Variadic template parameters. Waits just for <c>ELogConfig</c> instances.
+	/// </param>
+	/// <returns>
+	///		Reference to <c>this</c> object
+	///		to continue work with class methods in one line
+	/// </returns>
+	/// <example>
+	///		const auto* testBuilder = new CLogBuilder("TestName", ELogLevel::DEBUG);
+	///		auto* testLogger = testBuilder->Build();
+	///		testLogger->SetLogConfig(ELogConfig::MESSAGE, ELogConfig::LOG_LEVEL,
+	///		ELogConfig::PARAMS, ELogConfig::MESSAGE /* won't be added */);
+	/// </example>
+	template<typename... Args>
+	CLogger& CLogger::SetLogConfig(Args... args)
+	{
+		this->m_log_config_list.clear();
+		this->AddLogConfig(args...);
+		return *this;
+	}
+
+	/// <summary>
+	///		Prints to all streams, that <c>CLogger</c> stores, <c>CLogMessage</c> object
+	/// </summary>
+	/// <typeparam name="...Args">
+	///		Variadic template types of <c>CLogMessage</c>
+	///	</typeparam>
+	/// <param name="log_message">
+	///		<c>CLogMessage</c> to print
+	/// </param>
+	/// <example>
+	///		const auto* testBuilder = new CLogBuilder("TestName", ELogLevel::DEBUG);
+	///		auto* testLogger = testBuilder->Build();
+	///		testLogger->SetLogConfig(ELogConfig::MESSAGE, ELogConfig::LOG_LEVEL,
+	///		ELogConfig::PARAMS);
+	///		testLogger->PrintLogMessage(CLogMessage(/*...*/));
+	/// </example>
+	template<typename... Args>
+	void CLogger::PrintLogMessage(const CLogMessage<Args...>& log_message) const
+	{
+		if (this->m_log_level < log_message.GetLogLevel()
+			|| this->m_write_stream_list.empty()
+			|| this->m_log_config_list.empty())
+		{
+			return;
+		}
+
+		this->PrintToAllStreams(log_message);
+	}
+
+	/// <summary>
+	///		Prints some <c>CLogMessage</c> to all
+	///		output streams of <c>CLogger</c>
+	/// </summary>
+	/// <typeparam name="...Args">
+	///		
+	/// </typeparam>
+	/// <param name="log_message">
+	///		<c>CLogMessage</c> to print
+	/// </param>
+	template<typename ... Args>
+	void CLogger::PrintToAllStreams(const CLogMessage<Args...>& log_message) const
+	{
+		for (const auto& [stream, mutex] : this->m_write_stream_list)
+		{
+			if (nullptr != mutex)
+			{
+				mutex->lock();
+			}
+			
+			this->PrintLogMessage(log_message, stream);
+
+			if (nullptr != mutex)
```

Comment:
Do we need to check mutex 2 times on nullptr in one function?
Is possible that mutex might become nullptr somewhere during function run?

Reply:
Ok, I have idea how to rework this with `std::lock_guard`/`std::unique_lock`